packages/model/addon/-private/system/many-array.js

```diff
@@ -192,35 +207,14 @@ export default EmberObject.extend(MutableArray, DeprecatedEvented, {
     return internalModel.getRecord();
   },
 
-  flushCanonical(toSet, isInitialized = true) {
-    // It’s possible the parent side of the relationship may have been unloaded by this point
-    if (!_objectIsAlive(this)) {
-      return;
-    }
-    // diff to find changes
-    let diff = diffArray(this.currentState, toSet);
-    if (diff.firstChangeIndex !== null) {
-      // it's null if no change found
-      // we found a change
-      this.arrayContentWillChange(diff.firstChangeIndex, diff.removedCount, diff.addedCount);
-      this.set('length', toSet.length);
-      this.currentState = toSet.slice();
-      this.arrayContentDidChange(diff.firstChangeIndex, diff.removedCount, diff.addedCount);
-      if (isInitialized && diff.addedCount > 0) {
-        //notify only on additions
-        //TODO only notify if unloaded
-        this.internalModel.manyArrayRecordAdded(this.get('key'));
-      }
-    }
-  },
```

Comment:
This was the sync flush, right?

Reply:
This was part of it (called by retrieveLatest). The callback here was a weird backflow we were dependent on that I was able to refactor away. Removing this back flow is where most of this perf improvement is coming from in this PR since we still aren't lazy given pretty much everything is wrapped in an ArrayProxy still.

What the backflow was doing was ensuring unloaded internal models were cleared from relationships, fixing the relationship notification flow ensured this would be true without this.